index.qmd

```diff
@@ -0,0 +1,65 @@
+
+In the latest release of `arrow`, version 9.0.0, the [`show_exec_plan()`](https://arrow.apache.org/docs/r/reference/show_exec_plan.html) function is introduced. This function allows you to see the execution plan generated from your code. For example:
+
+```{r}
+#| echo: false
+#| output: false
+library(arrow)
+library(dplyr)
+```
+
+```{r}
+library(arrow)
+library(dplyr)
+mtcars %>%
+  arrow_table() %>%
+  filter(mpg > 20) %>%
+  mutate(wt_kg = (wt*1000) * 0.453592)  %>%
+  show_exec_plan()
+```
+
+This functionality is similar to that of [`dplyr::explain()`](https://dplyr.tidyverse.org/reference/explain.html) and you'll get the same result whether you call [`arrow::show_exec_plan()`](https://arrow.apache.org/docs/r/reference/show_exec_plan.html), [`dplyr::explain()`](https://dplyr.tidyverse.org/reference/explain.html) or [`dplyr::show_query()`](https://dplyr.tidyverse.org/reference/explain.html) on an `arrow_dplyr_query` object.
+
+## What is Acero?
+
+Acero is an Arrow-native query execution engine, developed as part of the Arrow C++ / R libraries.
+
+## What is a query execution engine?
+
+A query execution engine is a piece of software which allows users to execute queries against a data source. It typically has multiple steps, which could include things like taking the query and parsing it into an algebraic format, re-ordering and optimising the query in order to run it in the most efficient way, and actually running the query.
+
+## What is an ExecPlan?
+
+Queries to be run on Acero are specified as execution plans aka ExecPlans. These are directed graphs which express what operations need to take place. The graph for the ExecPlan in the code example earlier in this post looks like this.
+
+```{r}
+#| echo: false
+library(DiagrammeR)
+
+DiagrammeR::grViz("
+digraph graph2 {
+
+graph [layout = dot]
+
+# node definitions with substituted label text
+node [shape = rectangle, width = 4, fillcolor = Biege]
+a [label = 'TableSourceNode']
+b [label = 'FilterNode']
+c [label = 'ProjectNode']
+d [label = 'SinkNode']
+
+a -> b -> c -> d
+
+}
+
+")
+```
+
+The [TableSourceNode](https://arrow.apache.org/docs/cpp/streaming_execution.html#table-source) is used to input the data, which is then passed to the [FilterNode](https://arrow.apache.org/docs/cpp/streaming_execution.html#filter), where we apply our filter to only retain rows where the value of `mpg` is greater than 20. The [ProjectNode](https://arrow.apache.org/docs/cpp/streaming_execution.html#project) then specifies which columns we want, including the new column `wt_kg` as a [compute expression](https://arrow.apache.org/docs/cpp/api/compute.html). Finally, the [SinkNode](https://arrow.apache.org/docs/cpp/streaming_execution.html#sink) provides the output.
```

Comment:
You mentioned the first node is always a `source` and the last one always a `sink`. Should we mention that some operations are naturally sink nodes (e.g. `arrange()` is implemented with the `order_by_sink`)? They're naturally sink nodes, because, if we want to `arrange()` or take the first rows with `head()` we need to collect all the data in memory and then sort, etc.  

The implications for this is that anything following a sink node is part of a separate `ExecPlan`. This is how we can end up with nested or chained ExecPlans. We currently made the choice not to display these nested plans, but rather warn and return the input invisibly. 

With a single sink operation (input is a `TableSourceNode` and the output is an `OrderBySinkNode`: 
``` r
mtcars %>%
  arrow_table() %>%
  filter(mpg > 20) %>%
  mutate(x = gear/carb) %>%
  arrange(desc(hp)) %>% 
  show_exec_plan()
#> ExecPlan with 4 nodes:
#> 3:OrderBySinkNode{by={sort_keys=[FieldRef.Name(hp) DESC], null_placement=AtEnd}}
#>   2:ProjectNode{projection=[mpg, cyl, disp, hp, drat, wt, qsec, vs, am, gear, carb, "x": divide(cast(gear, {to_type=double, allow_int_overflow=false, allow_time_truncate=false, allow_time_overflow=false, allow_decimal_truncate=false, allow_float_truncate=false, allow_invalid_utf8=false}), cast(carb, {to_type=double, allow_int_overflow=false, allow_time_truncate=false, allow_time_overflow=false, allow_decimal_truncate=false, allow_float_truncate=false, allow_invalid_utf8=false}))]}
#>     1:FilterNode{filter=(mpg > 20)}
#>       0:TableSourceNode{}
```

if we add a call to `head()`, we get a warning:
``` r
mtcars %>%
  arrow_table() %>%
  filter(mpg > 20) %>%
  mutate(x = gear/carb) %>%
  arrange(desc(hp)) %>% 
  head() %>% 
  show_exec_plan()
#> Warning message:
#> The `ExecPlan` cannot be printed for a nested query.
```

We had to make a choice between printing the last `ExecPlan` (which would be confusing as it would have contained only 3 nodes: a `SourceNode`, a `ProjectNode` and a `SinkNode`) or printing the chained plans (which would involve actually running them, which is undesirable). 

Without the warning in place we would only print the plan from the point-of-view of the final ExecNode and would look like this (note how the input is no longer a `TableSourceNode`, but rather the end-point of the previous `ExecPlan`, i.e. a `SourceNode`):
``` r
mtcars %>%
  arrow_table() %>%
  filter(mpg > 20) %>%
  mutate(x = gear/carb) %>%
  arrange(desc(hp)) %>% 
  head() %>% 
  show_exec_plan()
#> ExecPlan with 3 nodes:
#> 2:SinkNode{}
#>   1:ProjectNode{projection=[mpg, cyl, disp, hp, drat, wt, qsec, vs, am, gear, carb, x]}
#>     0:SourceNode{}
```

Reply:
> You mentioned the first node is always a source and the last one always a sink

This is not quite what I've said, so I've rephrased it to be clear that I'm specifically talking about this example here and not in a general way.

>  Should we mention that some operations are naturally sink nodes (e.g. arrange() is implemented with the order_by_sink)

Yeah, sounds good to record this somewhere, but not here.  Do we want a new vignette containing this info and the rest of the things you mention in this comment?

Reply:
Or, we just add a section called "ExecPlans in Detail" and put it there.